### DISCUSSION TITLE
janet-good-repl: A Janet REPL with multi-line support and a persistent history

### DISCUSSION THREAD

**Discussion by MaxGyver83** [janet-lang/janet#1446] [Show and tell]
Hi all,

I would like to share a little project with you: <https://github.com/MaxGyver83/janet-good-repl>

> `janet-good-repl` is an improved REPL for the [Janet](https://janet-lang.org/) programming language. It's based on Janet's builtin REPL (version 1.34.0) and adds these features:
>
> * **Multi-line input**: Enter multiple lines that belong together (usually an s-expression) as a single history entry. This helps a lot when you want to repeat this expression (or a modified version of it) afterwards using the history. Use `Alt-Enter` to continue a multi-line entry.
> * **Bracketed paste**: Paste a code block (p.e. from an editor) as a single multi-line input.
> * **Persistent history**: The history is persistent between sessions.

Please test it and tell me if anything is unclear or doesn't work as expected. It definitely needs to be adapted to work on Windows.

Actually, my goal was to create real shell based on Janet (as a replacement for bash/zsh/fish etc). I think I'll create a separate thread for this.

**Comment by sogaiu** [comment]
Seems like some tweaking might be in order to get this working with a non-system install (e.g. for `PREFIX=$HOME/.local`) -- at least it didn't work for me as-is.

On a related note, [here](https://gist.github.com/sogaiu/e69f7696ecf3e5cffda9320697b17134) is a gist that uses some of spork's `getline` capabilities.

**Reply by MaxGyver83** [reply]
> Seems like some tweaking might be in order to get this working with a non-system install (e.g. for `PREFIX=$HOME/.local`) -- at least it didn't work for me as-is.

Thanks for the hint. I'll look into this!

> On a related note, [here](https://gist.github.com/sogaiu/e69f7696ecf3e5cffda9320697b17134) is a gist that uses some of spork's `getline` capabilities.

Thanks for sharing! I didn't know this yet. I'm not sure if it's working correctly. What should I see? So far I only noticed that it turns my home and my end key into backspace keys. Is this a `TERM` issue?

**Reply by sogaiu** [reply]
> What should I see?

I think some of the points of interest might be:

* [handing off of input via janet](https://github.com/janet-lang/spork/blob/5d52cb9d6a3756789c972bb091b387d15b9ac2a0/spork/getline.janet#L321-L383) (so not in C)
* [completion and docs related code also expressed in janet](https://github.com/janet-lang/spork/blob/5d52cb9d6a3756789c972bb091b387d15b9ac2a0/spork/getline.janet#L22-L59)
* [`getline` implementation that is non-blocking](https://github.com/janet-lang/spork/blob/5d52cb9d6a3756789c972bb091b387d15b9ac2a0/spork/getline.janet#L2)
* intention to work on multiple platforms including Windows (I think)
* written to be reused (vs janet's `shell.c`)

The machinery does use an underlying C module from spork - [rawterm](https://github.com/janet-lang/spork/blob/master/src/rawterm.c). 

> So far I only noticed that it turns my home and my end key into backspace keys. Is this a TERM issue?

Hmm, that doesn't happen for me.  Home and End keys go to the start and end of line respectively here.  I tested on Void Linux.

No idea if it's a `TERM` issue, sorry (^^;
